go/signedexchange/verifier.go

```diff
@@ -115,16 +116,47 @@ func (e *Exchange) Verify(verificationTime time.Time, certFetcher CertFetcher, l
 			l.Printf("validity-url (%s) is not same-origin with request URL (%v)", signature.ValidityUrl, e.RequestURI)
 			continue
 		}
+
 		// Step 2: "Use Section 3.5 to determine the signature's validity for
-		// requestUrl, headers, and payload, getting certificate-chain back. If
-		// this returned "invalid" or didn't return a certificate chain, return
-		// "invalid"."
+		//         requestUrl, headers, and payload, getting certificate-chain
+		//         back. If this returned "invalid" or didn't return a
+		//         certificate chain, return "invalid"."
 		_, err = verifySignature(e, verificationTime, certFetcher, signature)
 		if err != nil {
 			l.Printf("Verification of sinature %q failed: %v", signature.Label, err)
 			continue
 		}
-		// TODO: Implement Step 3-7.
+
+		// Step 3: "Let exchange be the exchange metadata and headers parsed out
+		//         of headers."
+		// e contains the exchange metadata and headers.
+
+		// Step 4: "If exchange's request method is not safe (Section 4.2.1 of
+		//         [RFC7231]) or not cacheable (Section 4.2.3 of [RFC7231]),
+		//         return "invalid"."
+		// Per [RFC7231], only GET and HEAD are safe and cacheable.
+		method := e.RequestHeaders.Get(":method")
+		if method != "GET" && method != "HEAD" {
+			l.Printf("Request method %q is not safe or not cacheable.", method)
+			continue
+		}
+
+		// Step 5: "If exchange's headers contain a stateful header field, as
+		//         defined in Section 4.1, return "invalid"."
+		for k := range e.RequestHeaders {
+			if IsStatefulRequestHeader(k) {
+				l.Printf("Exchange has stateful request header %q.", k)
+				continue SignatureLoop
+			}
+		}
```

Comment:
Would you factor out a helper method so we wont have label jump `continue`s?

Reply:
Done.